### PR TITLE
Feat - Add open and close tree events

### PR DESCRIPTION
**Pull request by HPRIOR**
Add 'open' and 'close' callbacks or 'hooks' which can be set during setup and which fire when toggling the sidebar. 

This can be used to interact with other plugins. For example, the position of a tab bar can be changed according to the width of the file tree when it is opened or closed.

### CODE REVIEW

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Please document in help and readme.

See CONTRIBUTING.md

File: lua/nvim-tree.lua [hunk 1]
```diff
@@ -414,6 +417,8 @@ local DEFAULT_OPTS = { -- BEGIN_DEFAULT_OPTS
       profile = false,
     },
   },
+  open_hook = function () end,
+  close_hook = function () end,
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:421]: Please default these to nil and run `update-default-opts.sh`.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

**Comment by alex-courtis** [comment]
I think @HPRIOR has started something big that will grow.

Let's name configs so that we can add to it. Perhaps:
```lua
  hooks = {
    node_open = nil,
    node_close = nil,
  },
```

Maybe `actions.hooks`.

@kyazdani42 you have name ideas ;)

**Review by alex-courtis — CHANGES_REQUESTED** [review]
These should be nvim-tree.events e.g. `nvim-tree.events.on_nvim_tree_open`

We could pass the node, like action callbacks.

**Comment by HPRIOR** [comment]
Hey @alex-courtis, I'll implement the suggested changes soon. 

With regard to your previous comment (I'm not sure where it has gone) an additional section in the config would be a good idea if 'hooks' are to be added elsewhere.  Something like:

```lua
hooks = {
    tree_open = nil,
    tree_close = nil,
},
```
Would it be possible to explain how the event system works?

**Review by kyazdani42 — COMMENTED** [review]
you don't need to setup the hooks yourself, the user will call the the event registration himself, no need to set it during setup :)

**Comment by kyazdani42** [comment]
i'm not even sure you placed the event at the right location to be honest :smile: i think i would place them in view.lua. There are multiple ways the tree can be opened, it won't just work by setting it in toggle :)

**Comment by HPRIOR** [comment]
@kyazdani42 thanks for your feedback. I have moved the event dispatches to `view.lua`. Let me know if there is anything else I can do. If the code is not to your liking otherwise, I would really like to see this feature at some point!

**Comment by alex-courtis** [comment]
> With regard to your previous comment an additional section in the config would be a good idea

That was a mistake on my part... before I realised that we should use events ;)

**Review by alex-courtis — CHANGES_REQUESTED** [review]
We need to document this in the help: nvim-tree.events

**Comment by HPRIOR** [comment]
> We need to document this in the help: nvim-tree.events

I've added the documentation

**Review by alex-courtis — APPROVED** [review]

**Review by kyazdani42 — APPROVED** [review]

**Comment by gegoune** [comment]
Another useful event would be `on_tree_leave` (or similar) for when window is not closed but no longer active.

**Comment by kyazdani42** [comment]
what do you mean no longer active ? when the tree buffer is replaced by a file buffer ?

**Comment by gegoune** [comment]
Apologies for not being clear. I mean event when cursor leaves nvim-tree window. Wither because file was opened or by any window movement, such as `<C-w><C-w>`.

**Comment by kyazdani42** [comment]
well...
```
au WinLeave NvimTree_* ...
```
?

**Comment by gegoune** [comment]
Absolutely, that's what I have at the moment, and with on open event I can replace one of autocmds, but only one. Leaving tree remains as autocmd rather than event, which feels bit odd.

**Comment by kyazdani42** [comment]
Well i'm not sure how it would be different inside the plugin ^^'

**Comment by gegoune** [comment]
That's absolutely alright, just thought that it could make events more complete. I am fine sticking to auto commands.

**Comment by kyazdani42** [comment]
:smile: now you can define autocmd in lua with callbacks so i think it doesn't matter too much :)

**Comment by gegoune** [comment]
And this is exactly what I have. Thanks though and apologies for the noise. :)

**Pull request merged**